packages/cli/src/services/public-api-key.service.ts

```diff
@@ -60,21 +73,76 @@ export class PublicApiKeyService {
 	/**
 	 * Redacts an API key by keeping the first few characters and replacing the rest with asterisks.
 	 * @param apiKey - The API key to be redacted. If null, the function returns undefined.
+	 * @param revealCount - Number of characters to reveal.
+	 * @param maxLength - Maximum length of the redacted API key.
 	 * @returns The redacted API key with a fixed prefix and asterisks replacing the rest of the characters.
 	 * @example
 	 * ```typescript
 	 * const redactedKey = PublicApiKeyService.redactApiKey('12345-abcdef-67890');
 	 * console.log(redactedKey); // Output: '12345-*****'
 	 * ```
 	 */
-	redactApiKey(apiKey: string) {
-		const keepLength = 5;
-		return (
-			API_KEY_PREFIX +
-			apiKey.slice(API_KEY_PREFIX.length, API_KEY_PREFIX.length + keepLength) +
-			'*'.repeat(apiKey.length - API_KEY_PREFIX.length - keepLength)
-		);
+	redactApiKey(apiKey: string, revealCount = 15, maxLength = 80) {
+		if (revealCount < 0 || revealCount > apiKey.length) {
+			throw new ApplicationError('Invalid reveal count');
+		}
+
+		const visiblePart = apiKey.slice(0, revealCount);
+		const redactedPart = '*'.repeat(apiKey.length - revealCount);
+
+		const completeRedactedApiKey = visiblePart + redactedPart;
+
+		return completeRedactedApiKey.slice(0, maxLength);
 	}
 
-	createApiKeyString = () => `${API_KEY_PREFIX}${randomBytes(40).toString('hex')}`;
+	private isLegacyApiKey(apiKey: string) {
+		return apiKey.startsWith(LEGACY_API_KEY_PREFIX);
+	}
+
+	private isValidApiKey(apiKey: string) {
+		try {
+			const { aud } = this.jwtService.verify(apiKey);
+			return aud === API_KEY_AUDIENCE;
+		} catch {
+			return false;
+		}
+	}
+
+	getAuthMiddleware(version: string) {
+		return async (
+			req: AuthenticatedRequest,
+			_scopes: unknown,
+			schema: OpenAPIV3.ApiKeySecurityScheme,
+		): Promise<boolean> => {
+			const providedApiKey = req.headers[schema.name.toLowerCase()] as string;
+
+			const user = await this.getUserForApiKey(providedApiKey);
+
+			if (!user) return false;
+
+			/*
+				Legacy API keys will be deprecated in n8n v2.
+				Then we can delete the first check and always validate the API key.
+			*/
+			if (!this.isLegacyApiKey(providedApiKey) && !this.isValidApiKey(providedApiKey)) {
+				return false;
+			}
```

Comment:
I wanted to validate the JWT audience to avoid people from using the JWT in the internal endpoints (both are signed using the same secret), but just realized that that would not work, as the internal endpoints auth middleware check for the hash and the public API JWT will never have that property set.